### PR TITLE
[FIRRTL] Cache a symbol table instead of doing linear lookups every instance.

### PR DESCRIPTION
**Pull request by darthscsi**
roughly triples modules parsing performance on large designs.  This is about a 2x speedup for overall parsing stage.

### CODE REVIEW

**Review by seldridge — APPROVED** [review]

**Comment by mikeurbach** [comment]
I just linked an issue I opened about this. I believe there are a couple other places we could do the lookup besides the one you've updated here. The issue had this search query: https://github.com/search?q=repo%3Allvm%2Fcirct%20path%3Alib%2FDialect%2FFIRRTL%2FImport%2FFIRParser.cpp%20%22lookupSymbol(%22&type=code

**Comment by mikeurbach** [comment]
I guess just one other, we look up optionGroup from the circuit's symbol table here: https://github.com/llvm/circt/blob/43e6f074db0de934f5411ace120080ca962c27fe/lib/Dialect/FIRRTL/Import/FIRParser.cpp#L3882-L3883

That should be able to use the cached symbol table.

The final use of lookupSymbol is within `optionGroup`, which is itself a symbol table, so we couldn't use the cache there, but it should also be a much smaller scan.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 1]
```diff
@@ -5453,10 +5459,17 @@ ParseResult FIRCircuitParser::parseCircuit(
   // proactively touch it to make sure that it is always already created.
   (void)getLexer().translateLocation(info.getFIRLoc());
 
+  // Pre-verify symbol table, so we can construct it next.  Ideally, we would do
+  // this verification through the trait.
+  if (failed(mlir::detail::verifySymbolTable(circuit)))
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:5464]: Little nervous about verifying symbol /users/ generically at this point, but probably it's fine.

It appears SymbolTable::lookup is actually "const", can we make the plumbed reference const as well? Might help ensure accidental insertion or mutation, hopefully that "just works"?

**Review by dtzSiFive — COMMENTED** [review]
LGTM, thanks for addressing this.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 1 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:5464]: users shouldn't even exist, but it is an extra scan.

**Pull request merged**